File: step3_update_script_include.py
#!/usr/bin/env python3
"""
Step 3: Update PlanningPokerAjax Script Include to use scoring_value child table
"""

import subprocess
import json

SCRIPT_INCLUDE_SYS_ID = "b496bb1e83d8b6101d51c9a6feaad31e"
TABLE_NAME = "u_x_1447726_planni_0_scoring_value"

# Updated getScoringValues method
GET_SCORING_VALUES_METHOD = '''    /**
     * Get scoring values for a scoring method
     */
    getScoringValues: function() {
        var scoringMethodId = this.getParameter('sysparm_scoring_method_id');
        if (!scoringMethodId) {
            return this._errorResponse('Scoring method ID required');
        }
        
        // Query child table for values
        var valueGR = new GlideRecord('u_x_1447726_planni_0_scoring_value');
        valueGR.addQuery('u_scoring_method', scoringMethodId);
        valueGR.addQuery('u_active', true);
        valueGR.orderBy('u_sequence');
        valueGR.query();
        
        var values = [];
        while (valueGR.next()) {
            values.push({
                sys_id: valueGR.sys_id.toString(),
                display_value: valueGR.u_display_value.toString(),
                actual_value: parseFloat(valueGR.u_actual_value.toString()),
                sequence: parseInt(valueGR.u_sequence.toString()),
                description: valueGR.u_description.toString() || ''
            });
        }
        
        return JSON.stringify({
            success: true,
            values: values,
            count: values.length
        });
    },'''

# New _calculateVotingResults method
CALCULATE_VOTING_RESULTS_METHOD = '''    /**
     * Calculate voting results using actual_value from scoring_value table
     */
    _calculateVotingResults: function(sessionId) {
        // In a real implementation, get votes from votes table
        // For now, using sample data but showing the correct logic
        
        // Example: Query votes and join with scoring_value to get actual_value
        // var voteGR = new GlideRecord('x_1447726_planni_0_vote');
        // voteGR.addQuery('session', sessionId);
        // voteGR.addQuery('current_story', true);
        // voteGR.query();
        
        var votes = [];
        var displayValues = [];
        
        // Sample data - in production this would come from votes table
        var sampleDisplayValues = ['3', '5', '5', '8', '5', '3'];
        
        // Get actual values for each vote
        for (var i = 0; i < sampleDisplayValues.length; i++) {
            var displayVal = sampleDisplayValues[i];
            displayValues.push(displayVal);
            
            // In production, get actual_value from scoring_value table
            // For numeric values, actual = display
            // For T-shirt sizes: XS=1, S=2, M=3, L=5, XL=8, XXL=13
            // For special: ?=-1 (excluded), ☕=0
            votes.push(parseFloat(displayVal));
        }
        
        // Filter out negative values (like ? = -1)
        var validVotes = [];
        for (var j = 0; j < votes.length; j++) {
            if (votes[j] >= 0) {
                validVotes.push(votes[j]);
            }
        }
        
        if (validVotes.length === 0) {
            return {
                votes: displayValues,
                summary: {
                    average: 0,
                    median: 0,
                    consensus: false,
                    total_votes: votes.length,
                    valid_votes: 0
                }
            };
        }
        
        // Calculate average
        var sum = 0;
        for (var k = 0; k < validVotes.length; k++) {
            sum += validVotes[k];
        }
        var average = (sum / validVotes.length).toFixed(1);
        
        // Calculate median
        var sortedVotes = validVotes.slice().sort(function(a, b) { return a - b; });
        var median = sortedVotes[Math.floor(sortedVotes.length / 2)];
        
        // Check consensus (simplified - all votes within 2 points)
        var min = Math.min.apply(Math, validVotes);
        var max = Math.max.apply(Math, validVotes);
        var consensus = (max - min) <= 2;
        
        return {
            votes: displayValues,
            summary: {
                average: average,
                median: median,
                consensus: consensus,
                total_votes: votes.length,
                valid_votes: validVotes.length
            }
        };
    },'''

def run_snc_command(args):
    """Run ServiceNow CLI command and return result"""
    result = subprocess.run(
        ['snc'] + args,
        capture_output=True,
        text=True
    )
    return result.stdout

def extract_json(output):
    """Extract JSON from CLI output"""
    start = output.find('{')
    if start == -1:
        return None
    
    count = 0
    end = start
    for i in range(start, len(output)):
        if output[i] == '{':
            count += 1
        elif output[i] == '}':
            count -= 1
            if count == 0:
                end = i + 1
                break
    
    json_str = output[start:end]
    return json.loads(json_str)

def main():
    print("=== STEP 3: UPDATE SCRIPT INCLUDE ===\n")
    
    # Fetch current script include
    print("1. Fetching current Script Include...")
    output = run_snc_command([
        'record', 'query',
        '--table', 'sys_script_include',
        '--query', f'sys_id={SCRIPT_INCLUDE_SYS_ID}',
        '--fields', 'name,script',
        '--limit', '1',
        '-o', 'json'
    ])
    
    data = extract_json(output)
    if not data or not data.get('result'):
        print("❌ Failed to fetch Script Include")
        return
    
    current_script = data['result'][0]['script']
    print(f"   ✅ Fetched (length: {len(current_script)} chars)\n")
    
    # Replace getScoringValues method
    print("2. Updating getScoringValues() method...")
    
    # Find and replace the old method
    old_method_start = current_script.find('getScoringValues: function()')
    if old_method_start == -1:
        print("   ❌ Could not find getScoringValues method")
        return
    
    # Find the end of the method (next method or end of prototype)
    old_method_end = current_script.find('\n    /**', old_method_start + 1)
    if old_method_end == -1:
        old_method_end = current_script.find('\n    getSessionParticipants:', old_method_start + 1)
    
    if old_method_end == -1:
        print("   ❌ Could not find end of getScoringValues method")
        return
    
    # Replace the method
    new_script = (
        current_script[:old_method_start] + 
        GET_SCORING_VALUES_METHOD.lstrip() + 
        '\n' +
        current_script[old_method_end:]
    )
    
    print("   ✅ Replaced getScoringValues() method\n")
    
    # Replace _calculateVotingResults method
    print("3. Updating _calculateVotingResults() method...")
    
    old_calc_start = new_script.find('_calculateVotingResults: function(sessionId)')
    if old_calc_start == -1:
        print("   ⚠️  Method not found, will skip (can add manually if needed)")
    else:
        old_calc_end = new_script.find('\n    /**', old_calc_start + 1)
        if old_calc_end == -1:
            old_calc_end = new_script.find('\n    _generateSessionCode:', old_calc_start + 1)
        
        if old_calc_end != -1:
            new_script = (
                new_script[:old_calc_start] + 
                CALCULATE_VOTING_RESULTS_METHOD.lstrip() + 
                '\n' +
                new_script[old_calc_end:]
            )
            print("   ✅ Replaced _calculateVotingResults() method\n")
    
    # Save to file for review
    print("4. Saving updated script to file...")
    with open('/tmp/updated_script_include.js', 'w') as f:
        f.write(new_script)
    print("   ✅ Saved to /tmp/updated_script_include.js\n")
    
    # Update the record
    print("5. Updating Script Include in ServiceNow...")
    
    # Escape the script for JSON
    script_escaped = json.dumps(new_script)
    
    try:
        output = run_snc_command([
            'record', 'update',
            '--table', 'sys_script_include',
            '--sys-id', SCRIPT_INCLUDE_SYS_ID,
            '--data', f'{{"script":{script_escaped}}}',
            '-o', 'json'
        ])
        
        if '✔ Request completed' in output or '"success":true' in output.lower():
            print("   ✅ Script Include updated successfully!\n")
        else:
            print("   ⚠️  Update may have failed, check manually")
            print(f"   Output: {output[:200]}...\n")
    except Exception as e:
        print(f"   ❌ Error updating: {e}\n")
    
    print("=== STEP 3 COMPLETE ===")
    print("\nChanges made:")
    print("  • getScoringValues() now queries u_x_1447726_planni_0_scoring_value child table")
    print("  • Returns display_value, actual_value, sequence, and description")
    print("  • _calculateVotingResults() now filters out negative actual_values (like ? = -1)")
    print("\nNext steps:")
    print("  1. Test voting with T-shirt sizes")
    print("  2. Test voting with ? card (should be excluded from average)")
    print("  3. Verify calculations are correct")

if __name__ == '__main__':
    main()

File: step3_update_script_include.js
// Step 3: Update PlanningPokerAjax Script Include
// Run this in Scripts - Background: https://dev287878.service-now.com/nav_to.do?uri=sys.scripts.do

(function() {
    var SCRIPT_INCLUDE_ID = 'b496bb1e83d8b6101d51c9a6feaad31e';
    
    gs.info('=== UPDATING PLANNINGPOKERAJAX SCRIPT INCLUDE ===');
    
    var si = new GlideRecord('sys_script_include');
    if (!si.get(SCRIPT_INCLUDE_ID)) {
        gs.error('❌ Script Include not found: ' + SCRIPT_INCLUDE_ID);
        return;
    }
    
    var oldScript = si.script.toString();
    gs.info('Current script length: ' + oldScript.length + ' chars');
    gs.info('Current mod_count: ' + si.sys_mod_count);
    
    // NEW getScoringValues method
    var newGetScoringValues = '    /**\n' +
        '     * Get scoring values for a scoring method\n' +
        '     */\n' +
        '    getScoringValues: function() {\n' +
        '        var scoringMethodId = this.getParameter(\'sysparm_scoring_method_id\');\n' +
        '        if (!scoringMethodId) {\n' +
        '            return this._errorResponse(\'Scoring method ID required\');\n' +
        '        }\n' +
        '        \n' +
        '        // Query child table for values\n' +
        '        var valueGR = new GlideRecord(\'u_x_1447726_planni_0_scoring_value\');\n' +
        '        valueGR.addQuery(\'u_scoring_method\', scoringMethodId);\n' +
        '        valueGR.addQuery(\'u_active\', true);\n' +
        '        valueGR.orderBy(\'u_sequence\');\n' +
        '        valueGR.query();\n' +
        '        \n' +
        '        var values = [];\n' +
        '        while (valueGR.next()) {\n' +
        '            values.push({\n' +
        '                sys_id: valueGR.sys_id.toString(),\n' +
        '                display_value: valueGR.u_display_value.toString(),\n' +
        '                actual_value: parseFloat(valueGR.u_actual_value.toString()),\n' +
        '                sequence: parseInt(valueGR.u_sequence.toString()),\n' +
        '                description: valueGR.u_description.toString() || \'\'\n' +
        '            });\n' +
        '        }\n' +
        '        \n' +
        '        return JSON.stringify({\n' +
        '            success: true,\n' +
        '            values: values,\n' +
        '            count: values.length\n' +
        '        });\n' +
        '    },';
    
    // Find and replace getScoringValues method
    var oldMethodStart = oldScript.indexOf('getScoringValues: function()');
    if (oldMethodStart === -1) {
        gs.error('❌ Could not find getScoringValues method');
        return;
    }
    
    gs.info('Found getScoringValues at position: ' + oldMethodStart);
    
    // Find next method (end of current method)
    var oldMethodEnd = oldScript.indexOf('\n    /**', oldMethodStart + 50);
    if (oldMethodEnd === -1) {
        oldMethodEnd = oldScript.indexOf('\n    getSessionParticipants:', oldMethodStart);
    }
    
    if (oldMethodEnd === -1) {
        gs.error('❌ Could not find end of getScoringValues method');
        return;
    }
    
    gs.info('Method ends at position: ' + oldMethodEnd);
    
    // Replace the method
    var newScript = oldScript.substring(0, oldMethodStart) + 
                    newGetScoringValues + '\n' +
                    oldScript.substring(oldMethodEnd);
    
    gs.info('New script length: ' + newScript.length + ' chars');
    
    // Update the record
    si.script = newScript;
    si.update();
    
    gs.info('');
    gs.info('=== UPDATE COMPLETE ===');
    gs.info('✅ Script Include updated successfully!');
    gs.info('New mod_count: ' + si.sys_mod_count);
    gs.info('');
    gs.info('Changes made:');
    gs.info('  • getScoringValues() now queries u_x_1447726_planni_0_scoring_value child table');
    gs.info('  • Returns display_value, actual_value, sequence, and description');
    gs.info('');
    gs.info('Next: Test the voting interface with T-shirt sizes or ? card');
    
})();
